source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -182,1257 +182,1274 @@ namespace
-                    prependXInclude(doc, uri);
-                }
+                PvtPrim* parentLook = findPrimOrThrow(RtIdentifier(inheritString), parent, mapper);
+                RtLook rtLook(childLook->hnd());
+                rtLook.getInherit().connect(parentLook->hnd());
             }
         }
     }
 
-    void writePrimData(DocumentPtr& doc, const RtPrim& prim, const RtWriteOptions* options)
+    // Read in a look group. This assumes that all referenced looks have
+    // already been created.
+    PvtPrim* readLookGroup(const LookGroupPtr& src, PvtStage* stage, PvtRenamingMapper& mapper)
     {
-        const PvtPrim* p = PvtObject::cast<PvtPrim>(prim);
-        const RtIdentifier typeName = prim.getTypeInfo()->getShortTypeName();
-        if (p->hasApi<RtNodeDef>())
-        {
-            writeNodeDef(p, doc, options);
-        }
-        else if (p->hasApi<RtNode>())
+        const string LIST_SEPARATOR(",");
+
+        PvtPrim* parent = stage->getRootPrim();
+        const RtIdentifier name(src->getName());
+        PvtPrim* prim = stage->createPrim(parent->getPath(), name, RtLookGroup::typeName());
+        mapper.addMapping(parent, name, prim->getName());
+
+        RtLookGroup lookGroup(prim->hnd());
+        lookGroup.setEnabledLooks(src->getEnabledLooksString());
+
+        // Link to looks
+        const string& lookNamesString = src->getLooks();
+        StringVec lookNamesList  = splitString(lookNamesString, LIST_SEPARATOR);
+        for (auto lookName : lookNamesList)
         {
-            if (p->hasApi<RtNodeGraph>())
-            {
-                writeNodeGraph(p, doc, options);
-            }
-            else
+            if (!lookName.empty())
             {
-                writeNode(p, doc, options);
+                PvtPrim* lookPrim = findPrimOrThrow(RtIdentifier(lookName), parent, mapper);
+                lookGroup.addLook(lookPrim->hnd());
             }
         }
-        else if (p->hasApi<RtNodeImpl>())
-        {
-            writeImplementation(p, doc, options);
-        }
-        else if (p->hasApi<RtTargetDef>())
+
+        readAttributes(src, prim, lookGroup.getPrimSpec(), lookGroupIgnoreList);
+
+        return prim;
+    }
+
+    // Read in all look information from a document. Collections, looks and
+    // look groups are read in first. Then relationship linkages are made.
+    void readLookInformation(const DocumentPtr& doc, PvtStage* stage, const RtReadOptions* options, PvtRenamingMapper& mapper)
+    {
+        // Create the mergedDoc
+        DocumentPtr mergedDoc;
+        if (options && options->lookModifier)
         {
-            writeTargetDef(p, doc, options);
+            // Write the current runtime contents to the currentDoc
+            DocumentPtr currentDoc = createDocument();
+            RtWriteOptions wops;
+            writeDocument(currentDoc, stage, &wops);
+
+            mergedDoc = createDocument();
+            options->lookModifier(currentDoc, doc, mergedDoc);
```

Comment:
Suggestion:
1. We pass the stage to the lookModifier(stage, doc); 
2. The lookModifier returns the new document.
Then concept of merging and creating extra documents is hidden as part of the modifier.

i.e. the interface is:

using LookModifier =  std::function<DocumentPtr(const StagePtr state, const DocumentPtr& loadedDoc,)>;

Reply:
I'm afraid this doesn't work because writeDocument isn't an exposed interface and would need to be called from lookModifier in this case. That's why the documents need to be created in readLookInformation.